gocourse06/smartzoo/main.go

```diff
@@ -0,0 +1,176 @@
+Уникнення «Race Conditions»: забезпечте, щоб спільні ресурси (наприклад, дані про стан тварин) були захищені від
+одночасного доступу декількома горутинами. Використовуйте канали для синхронізації доступу.
+Управління «Deadlocks»: уважно використовуйте блокування та канали, щоб уникнути взаємних блокувань між горутинами.
+Логування та моніторинг: реалізуйте систему логування, яка фіксує важливі події у системі, наприклад, коли тварина
+потребує уваги або коли кормушка порожня.
+Тестування: напишіть модульні тести для перевірки коректності взаємодії між горутинами та уникнення «race conditions»
+та «deadlocks».
+*/
+package main
+
+import (
+	"fmt"
+	"log/slog"
+	"math/rand/v2"
+	"os"
+	"sync"
+	"time"
+)
+
+// Animal описує структуру даних для тварини
+type Animal struct {
+	ID     int
+	Health int
+	Hunger int
+	Mood   int
+}
+
+// Enclosure описує статус вольєра
+type Enclosure struct {
+	ID     int
+	IsOpen bool
+}
+
+// Feeder описує статус кормушки
+type Feeder struct {
+	ID      int
+	IsEmpty bool
+}
+
+// Генерує тестові дані для тварин
+func generateAnimals(n int) []Animal {
+	animals := make([]Animal, n)
+	for i := range n {
+		animals[i] = Animal{
+			ID:     i,
+			Health: rand.IntN(100),
+			Hunger: rand.IntN(100),
+			Mood:   rand.IntN(100),
+		}
+	}
+	return animals
+}
+
+// Генерує тестові дані для вольєрів
+func generateEnclosures(n int) []Enclosure {
+	enclosures := make([]Enclosure, n)
+	for i := range n {
+		enclosures[i] = Enclosure{
+			ID:     i,
+			IsOpen: rand.IntN(2) == 1,
+		}
+	}
+	return enclosures
+}
+
+// Генерує тестові дані для годівниці
+func generateFeeders(n int) []Feeder {
+	feeders := make([]Feeder, n)
+	for i := range n {
+		feeders[i] = Feeder{
+			ID:      i,
+			IsEmpty: rand.IntN(2) == 1,
+		}
+	}
+	return feeders
+}
+
+func controlCondition(animalChan <-chan *Animal, wg *sync.WaitGroup, log *slog.Logger) {
+	defer wg.Done()
+	time.Sleep(1 * time.Second)
+
+	if animal, ok := <-animalChan; ok {
+		if animal.Health < 30 {
+			log.Warn("Animal needs help", slog.Int("animalId", animal.ID))
+		}
+		if animal.Hunger < 30 {
+			log.Warn("Animal needs to be fed", slog.Int("animalId", animal.ID))
+		}
+		if animal.Mood < 30 {
+			log.Warn("Animal - needs to be released from the cage", slog.Int("animalId", animal.ID))
+		}
+	}
+}
+
+func controlEnclosure(enclosureChan chan *Enclosure, isOpen bool, wg *sync.WaitGroup, log *slog.Logger) {
+	defer wg.Done()
+	if enclosure, ok := <-enclosureChan; ok {
+		enclosure.IsOpen = isOpen
+		if isOpen {
+			log.Info("Closing the enclosure...")
+		} else {
+			log.Info("Opening the enclosure...")
+		}
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func controlFeeder(feedChan chan *Feeder, wg *sync.WaitGroup, log *slog.Logger) {
+	defer wg.Done()
+	time.Sleep(3 * time.Second) // must be more than 3
+	if feed, ok := <-feedChan; ok {
+		if feed.IsEmpty {
+			log.Warn("Needs a refill feeder", slog.Int("feedId", feed.ID))
+		}
+	}
+}
+
+func loggerNew(writer *os.File) *slog.Logger {
```

Comment:
Ось так має працювати, тому що `os.File` імплементує `io.Writer`:
```suggestion
func loggerNew(writer io.Writer) *slog.Logger {
```

Насправді, методу [`slog.NewJSONHandler`](https://pkg.go.dev/log/slog#NewJSONHandler) потрібен лише щось, що імплементує `io.Writer`:

```go
// NewJSONHandler creates a [JSONHandler] that writes to w,
// using the given options.
// If opts is nil, the default options are used.
func NewJSONHandler(w io.Writer, opts *HandlerOptions) *JSONHandler {
```

Reply:
Так більш канонічно:
```suggestion
func newLogger(writer io.Writer) *slog.Logger {
```